### PR TITLE
SSRF in - Cairosvg

### PR DESCRIPTION
**Pull request by b1nslashsh**
# :bookmark_tabs: Description

 `CairoSVG` is an SVG converter based on Cairo. It can export SVG files to PDF, EPS, PS, and PNG files. **useing this package in webapps can be lead to** `SSRF` **attacks**

Reference :-
**[Github](https://github.com/Kozea/CairoSVG)
[Pypi](https://pypi.org/project/CairoSVG/)
[cairosvg.org](https://cairosvg.org/documentation/)**
# :recycle:  Steps To Reproduce-:  
  0) `$ pip3 install cairosvg`
  1) `$ cairosvg payload.svg`
  1) creat a payload svg or use : [rc3.svg](https://drive.google.com/file/d/1jGhUXepvOV9bs_aaCThSmqloY_C1Nsj4/view?usp=sharing)

  rc3.svg
  ```
  <?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" style="overflow: hidden; position: relative;" width="300" height="200">
<image x="10" y="10" width="276" height="110" xlink:href="http://localhost:8080/svg" stroke-width="1" id="image3204" />
<rect x="0" y="150" height="10" width="300" style="fill: black"/>
</svg>

  ```
# :scientist: POC :-
poc

[gdrive](https://drive.google.com/file/d/1VqfAgldmtY-qrgHRfizVvAH2oIFZVT-q/view?usp=sharing)
## 💥 Impact
**SSRF  or Server-side request forgery**

### CODE REVIEW

**Review by Mik317 — COMMENTED** [review]
The fact the `unsafe` argument isn't necessary makes it a worst scenario but looking at the code I'm not sure this is out of their intentions: https://github.com/Kozea/CairoSVG/blob/f0edc6efd725aaf654bf1acbfc487d948941f191/cairosvg/url.py#L140 (they also return content of files in case a special url is provided ... mostly something wanted).

Not sure if this is by-design and therefore the only way to avoid this is just not use this tool or if the `unsafe` argument could also be used to identify a proper way to handle this.

I'd like to know what other sheriffs think about it :smile: @bbeale @mufeedvh :+1: 

Cheers,
Mik

**Comment by mufeedvh** [comment]
Hey @b1nslashsh,

This seems like a design choice to me as well but this can be fixed for the schemed URL handling part (https://github.com/Kozea/CairoSVG/blob/f0edc6efd725aaf654bf1acbfc487d948941f191/cairosvg/url.py#L138) where you can validate the given URL with a blacklist like `127.0.0.1, 0.0.0.0, [::], etc` but for the `file://` part as @Mik317 already stated, that in and of itself can be used as an SSRF (enumeration). Although I think this can be handled or improved upon by using the `unsafe` option/feature to use a blacklist before passing in the URL, in that case this can be considered valid.

In short, using a blacklist might be able to reduce attack/enumeration scope and only allowing what's intended to be given as input, for example, a valid SVG file processed after parsing. :+1: 

What do you guys think? @Mik317 @bbeale @toufik-airane @mzfr

**Review by mufeedvh — COMMENTED** [review]
Hey @b1nslashsh,

This seems like a design choice to me as well but this can be fixed for the schemed URL handling part (https://github.com/Kozea/CairoSVG/blob/f0edc6efd725aaf654bf1acbfc487d948941f191/cairosvg/url.py#L138) where you can validate the given URL with a blacklist like `127.0.0.1, 0.0.0.0, [::], etc` but for the `file://` part as @Mik317 already stated, that in and of itself can be used as an SSRF (enumeration). Although I think this can be handled or improved upon by using the `unsafe` option/feature to use a blacklist before passing in the URL, in that case this can be considered valid.

In short, using a blacklist might be able to reduce attack/enumeration scope and only allowing what's intended to be given as input, for example, a valid SVG file processed after parsing. :+1: 

What do you guys think? @Mik317 @bbeale @toufik-airane @mzfr

**Comment by Mik317** [comment]
Yeah, I agree with @mufeedvh :+1: 
We could introduce a blacklisting approach (we'll have to check for eventual bypasses in fix) when the `unsafe` option isn't applied :100: 

Cheers,
Mik

**Comment by roo7break** [comment]
Hello all, could you please update the vulnerable version list since Cairosvg implemented fix for this vulnerability when they release 2.7.0 on March 20, 2023. 

https://github.com/Kozea/CairoSVG/blob/main/NEWS.rst

Thanks in advance
